File: ios/MyNoteCore/Sources/MyNoteCore/APIClient.swift
import Foundation

public enum APIError: Error, Equatable, Sendable {
    case unauthenticated
    /// The account has no active `cloud_sync` entitlement — show the paywall.
    case subscriptionRequired
    case quotaExceeded(message: String)
    case conflict(message: String)
    case server(status: Int, code: String, message: String)
    case offline
    case decoding(String)
}

public struct Entitlement: Codable, Equatable, Sendable {
    public var entitlement: String
    public var active: Bool
    public var status: String
    public var expiresAt: Int64?
    public var productId: String
    public var platform: String
}

public struct EntitlementsResponse: Codable, Sendable {
    public var entitlements: [Entitlement]
}

/// The subset of the API the sync engine depends on. Extracted so the engine
/// can be driven by a scripted fake in tests instead of a live Worker.
public protocol SyncAPI: Sendable {
    func sync(cursor: Int, changes: [Change], limit: Int?) async throws -> SyncResponse
}

/// Talks to the Cloudflare Worker. The Firebase ID token is fetched lazily per
/// request so a refreshed token is picked up without recreating the client.
public actor APIClient: SyncAPI {
    private let baseURL: URL
    private let session: URLSession
    private let tokenProvider: @Sendable () async throws -> String?

    public init(baseURL: URL,
                session: URLSession = .shared,
                tokenProvider: @escaping @Sendable () async throws -> String?) {
        self.baseURL = baseURL
        self.session = session
        self.tokenProvider = tokenProvider
    }

    public func sync(cursor: Int, changes: [Change], limit: Int? = nil) async throws -> SyncResponse {
        try await post("/v1/sync", body: SyncRequest(cursor: cursor, changes: changes, limit: limit))
    }

    public func entitlements() async throws -> [Entitlement] {
        let res: EntitlementsResponse = try await get("/v1/entitlements")
        return res.entitlements
    }

    public func verifyAppleTransaction(id: String) async throws -> [Entitlement] {
        struct Body: Codable { let transactionId: String }
        let res: EntitlementsResponse = try await post("/v1/iap/apple/verify", body: Body(transactionId: id))
        return res.entitlements
    }

    // MARK: - Transport

    private func get<Response: Decodable>(_ path: String) async throws -> Response {
        try await send(request(path, method: "GET"))
    }

    private func post<Body: Encodable, Response: Decodable>(_ path: String, body: Body) async throws -> Response {
        var req = request(path, method: "POST")
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.httpBody = try JSONEncoder().encode(body)
        return try await send(req)
    }

    private func request(_ path: String, method: String) -> URLRequest {
        var req = URLRequest(url: baseURL.appendingPathComponent(path))
        req.httpMethod = method
        // Sync must never block the UI; a stalled network is just "try later".
        req.timeoutInterval = 30
        return req
    }

    private func send<Response: Decodable>(_ base: URLRequest) async throws -> Response {
        var req = base
        guard let token = try await tokenProvider() else { throw APIError.unauthenticated }
        req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")

        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await session.data(for: req)
        } catch {
            throw APIError.offline
        }

        guard let http = response as? HTTPURLResponse else {
            throw APIError.decoding("not an HTTP response")
        }

        if http.statusCode >= 400 {
            throw Self.mapError(status: http.statusCode, data: data)
        }

        do {
            return try JSONDecoder().decode(Response.self, from: data)
        } catch {
            throw APIError.decoding(String(describing: error))
        }
    }

    struct ErrorBody: Decodable { let error: String?; let code: String? }

    static func mapError(status: Int, data: Data) -> APIError {
        let body = try? JSONDecoder().decode(ErrorBody.self, from: data)
        let code = body?.code ?? "error"
        let message = body?.error ?? "request failed"

        switch (status, code) {
        case (401, _):               return .unauthenticated
        case (402, _):               return .subscriptionRequired
        case (507, _):               return .quotaExceeded(message: message)
        case (409, _):               return .conflict(message: message)
        default:                     return .server(status: status, code: code, message: message)
        }
    }
}

File: android/core/src/main/kotlin/io/mynote/core/ApiClient.kt
package io.mynote.core

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.Serializable
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.IOException
import java.util.concurrent.TimeUnit

@Serializable
data class Entitlement(
    val entitlement: String,
    val active: Boolean,
    val status: String,
    val expiresAt: Long? = null,
    val productId: String,
    val platform: String,
)

@Serializable
data class EntitlementsResponse(val entitlements: List<Entitlement>)

@Serializable
private data class ErrorBody(val error: String? = null, val code: String? = null)

@Serializable
private data class GoogleVerifyRequest(val productId: String, val purchaseToken: String)

/**
 * Talks to the Cloudflare Worker.
 *
 * The Firebase ID token is fetched per request via [tokenProvider] so a refreshed
 * token is picked up without rebuilding the client.
 */
class ApiClient(
    private val baseUrl: String,
    private val tokenProvider: suspend () -> String?,
    private val client: OkHttpClient = defaultClient(),
) : SyncApi {

    override suspend fun sync(cursor: Int, changes: List<Change>, limit: Int?): SyncResponse =
        post("/v1/sync", MyNoteJson.encodeToString(SyncRequest.serializer(), SyncRequest(cursor, changes, limit)))
            .let { MyNoteJson.decodeFromString(SyncResponse.serializer(), it) }

    suspend fun entitlements(): List<Entitlement> =
        MyNoteJson.decodeFromString(EntitlementsResponse.serializer(), get("/v1/entitlements")).entitlements

    suspend fun verifyGooglePurchase(productId: String, purchaseToken: String): List<Entitlement> {
        val body = MyNoteJson.encodeToString(
            GoogleVerifyRequest.serializer(),
            GoogleVerifyRequest(productId, purchaseToken),
        )
        return MyNoteJson
            .decodeFromString(EntitlementsResponse.serializer(), post("/v1/iap/google/verify", body))
            .entitlements
    }

    // ---- transport ---------------------------------------------------------

    private suspend fun get(path: String): String = execute(requestBuilder(path).get())

    private suspend fun post(path: String, json: String): String =
        execute(requestBuilder(path).post(json.toRequestBody(JSON_MEDIA)))

    private suspend fun requestBuilderToken(): String {
        return tokenProvider() ?: throw ApiException(ApiError.Unauthenticated)
    }

    private fun requestBuilder(path: String): Request.Builder =
        Request.Builder().url(baseUrl.trimEnd('/') + path)

    private suspend fun execute(builder: Request.Builder): String = withContext(Dispatchers.IO) {
        val request = builder
            .header("Authorization", "Bearer ${requestBuilderToken()}")
            .header("Content-Type", "application/json")
            .build()

        val response = try {
            client.newCall(request).execute()
        } catch (e: IOException) {
            throw ApiException(ApiError.Offline)
        }

        response.use {
            val body = it.body?.string().orEmpty()
            if (!it.isSuccessful) throw ApiException(mapError(it.code, body))
            body
        }
    }

    private fun mapError(status: Int, body: String): ApiError {
        val parsed = runCatching { MyNoteJson.decodeFromString(ErrorBody.serializer(), body) }.getOrNull()
        val message = parsed?.error ?: "request failed"
        return when (status) {
            401 -> ApiError.Unauthenticated
            402 -> ApiError.SubscriptionRequired
            507 -> ApiError.QuotaExceeded(message)
            409 -> ApiError.Conflict(message)
            else -> ApiError.Server(status, parsed?.code ?: "error", message)
        }
    }

    companion object {
        private val JSON_MEDIA = "application/json; charset=utf-8".toMediaType()

        /** Sync must never block the UI, so a stalled network just means "later". */
        fun defaultClient(): OkHttpClient = OkHttpClient.Builder()
            .connectTimeout(15, TimeUnit.SECONDS)
            .readTimeout(30, TimeUnit.SECONDS)
            .writeTimeout(30, TimeUnit.SECONDS)
            .retryOnConnectionFailure(true)
            .build()
    }
}
